File: video/splash.c
/*
 * SlopOS Splash Screen Implementation
 * Displays boot splash screen with logo and loading progress
 */

#include <stdint.h>
#include <stddef.h>
#include "splash.h"
#include "framebuffer.h"
#include "graphics.h"
#include "font.h"
#include "../drivers/serial.h"
#include "../drivers/pit.h"
#include "../lib/klog.h"

/* ========================================================================
 * SPLASH SCREEN IMPLEMENTATION
 * ======================================================================== */

/*
 * Draw SlopOS logo as ASCII art using graphics primitives
 */
static int splash_draw_logo(int center_x, int center_y) {
    if (!framebuffer_is_initialized()) {
        return -1;
    }

    // Calculate logo dimensions and position
    int logo_width = 300;
    int logo_height = 150;
    int logo_x = center_x - logo_width / 2;
    int logo_y = center_y - logo_height / 2;

    // Draw main logo rectangle with gradient effect
    for (int y = 0; y < logo_height; y++) {
        uint32_t gradient_intensity = 0x40 + (y * 0x80 / logo_height);
        uint32_t gradient_color = (gradient_intensity << 24) | (gradient_intensity << 16) | 0xFF;
        graphics_draw_hline(logo_x, logo_x + logo_width, logo_y + y, gradient_color);
    }

    // Draw logo border
    graphics_draw_rect(logo_x - 2, logo_y - 2, logo_width + 4, logo_height + 4, SPLASH_LOGO_COLOR);

    // Draw stylized "SLOP" letters using geometric shapes
    int letter_spacing = 60;
    int letter_start_x = logo_x + 30;
    int letter_y = logo_y + 40;
    int letter_height = 70;

    // Letter S - curves approximated with rectangles
    graphics_draw_rect_filled(letter_start_x, letter_y, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y + 25, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y + 55, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y, 15, 40, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x + 25, letter_y + 30, 15, 40, SPLASH_LOGO_COLOR);

    // Letter L
    letter_start_x += letter_spacing;
    graphics_draw_rect_filled(letter_start_x, letter_y, 15, letter_height, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y + letter_height - 15, 40, 15, SPLASH_LOGO_COLOR);

    // Letter O
    letter_start_x += letter_spacing;
    graphics_draw_rect_filled(letter_start_x, letter_y, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y + letter_height - 15, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y, 15, letter_height, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x + 25, letter_y, 15, letter_height, SPLASH_LOGO_COLOR);

    // Letter P
    letter_start_x += letter_spacing;
    graphics_draw_rect_filled(letter_start_x, letter_y, 15, letter_height, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x, letter_y + 25, 40, 15, SPLASH_LOGO_COLOR);
    graphics_draw_rect_filled(letter_start_x + 25, letter_y, 15, 40, SPLASH_LOGO_COLOR);

    return 0;
}

/*
 * Draw progress bar
 */
static int splash_draw_progress_bar(int x, int y, int width, int height, int progress) {
    if (!framebuffer_is_initialized()) {
        return -1;
    }

    // Draw progress bar background
    graphics_draw_rect_filled(x, y, width, height, 0x333333FF);

    // Draw progress bar border
    graphics_draw_rect(x - 1, y - 1, width + 2, height + 2, SPLASH_LOGO_COLOR);

    // Draw progress fill
    if (progress > 0) {
        int fill_width = (width * progress) / 100;
        graphics_draw_rect_filled(x, y, fill_width, height, SPLASH_PROGRESS_COLOR);
    }

    return 0;
}

// Global splash screen state
static int splash_active = 0;
static int current_progress = 0;

/*
 * Initialize splash screen (without fake animation)
 */
int splash_show_boot_screen(void) {
    if (!framebuffer_is_initialized()) {
        klog_printf(KLOG_INFO, "SPLASH: Framebuffer not initialized\n");
        return -1;
    }

    klog_printf(KLOG_INFO, "SPLASH: Displaying boot splash screen...\n");

    // Clear screen with splash background color
    framebuffer_clear(SPLASH_BG_COLOR);

    // Get screen dimensions
    uint32_t width = framebuffer_get_width();
    uint32_t height = framebuffer_get_height();
    int center_x = width / 2;
    int center_y = height / 2;

    // Draw logo
    splash_draw_logo(center_x, center_y - 80);

    // Draw title text
    font_draw_string(center_x - 80, center_y + 100, "SlopOS v0.000069", SPLASH_TEXT_COLOR, 0x00000000);
    font_draw_string(center_x - 120, center_y + 120, "the ultimate vibe slop experience", SPLASH_TEXT_COLOR, 0x00000000);

    // Draw loading message
    font_draw_string(center_x - 40, center_y + 160, "Initializing...", SPLASH_TEXT_COLOR, 0x00000000);

    // Draw initial progress bar at 0%
    int progress_bar_width = 300;
    int progress_bar_height = 20;
    int progress_bar_x = center_x - progress_bar_width / 2;
    int progress_bar_y = center_y + 200;

    splash_draw_progress_bar(progress_bar_x, progress_bar_y, progress_bar_width, progress_bar_height, 0);

    // Mark splash as active and reset progress
    splash_active = 1;
    current_progress = 0;

    klog_printf(KLOG_INFO, "SPLASH: Boot splash screen initialized\n");

    // No initial delay - let the boot process drive the timing

    return 0;
}

/*
 * Report progress during boot (called from kernel initialization)
 */
int splash_report_progress(int progress, const char *message) {
    if (!splash_active || !framebuffer_is_initialized()) {
        return -1;
    }

    // Update progress
    current_progress = progress;
    if (current_progress > 100) current_progress = 100;

    klog_printf(KLOG_INFO, "SPLASH: Progress %d%% - %s\n",
                current_progress,
                message ? message : "...");

    // Update the visual progress bar and message
    int result = splash_update_progress(current_progress, message);

    // Add brief delays between steps for 4 second total boot time
    // With 14 total steps, each delay should be ~285ms for 4 second total
    uint32_t delay_ms = 280; // Base delay - 0.28 seconds

    // Slight variation based on operation type for realistic feel
    if (current_progress <= 20) {
        delay_ms = 300; // Graphics initialization - 0.3 seconds
    } else if (current_progress <= 40) {
        delay_ms = 250; // Early system setup - 0.25 seconds
    } else if (current_progress <= 60) {
        delay_ms = 280; // APIC/interrupt setup - 0.28 seconds
    } else if (current_progress <= 80) {
        delay_ms = 320; // PCI enumeration takes longer - 0.32 seconds
    } else if (current_progress <= 95) {
        delay_ms = 280; // Scheduler/task setup - 0.28 seconds
    } else {
        delay_ms = 250; // Final completion - 0.25 seconds
    }

    // Apply the delay
    pit_poll_delay_ms(delay_ms);

    return result;
}

/*
 * Draw the post-boot graphics demo screen
 * Shows colored rectangles, circle, border, and welcome text
 */
int splash_draw_graphics_demo(void) {
    if (!framebuffer_is_initialized()) {
        return -1;
    }

    // Clear splash screen and show graphics demo (like in 8fe117b)
    framebuffer_clear(0x001122FF);

    // Initialize console with white text on dark background
    font_console_init(0xFFFFFFFF, 0x00000000);

    // Draw graphics demo
    graphics_draw_rect_filled(20, 20, 300, 150, 0xFF0000FF);        // Red rectangle
    graphics_draw_rect_filled(700, 20, 300, 150, 0x00FF00FF);       // Green rectangle
    graphics_draw_circle(512, 384, 100, 0xFFFF00FF);                // Yellow circle

    // White border around entire screen
    graphics_draw_rect_filled(0, 0, 1024, 4, 0xFFFFFFFF);           // Top
    graphics_draw_rect_filled(0, 764, 1024, 4, 0xFFFFFFFF);         // Bottom
    graphics_draw_rect_filled(0, 0, 4, 768, 0xFFFFFFFF);            // Left
    graphics_draw_rect_filled(1020, 0, 4, 768, 0xFFFFFFFF);         // Right

    // Display welcome message using font_draw_string
    font_draw_string(20, 600, "*** SLOPOS GRAPHICS SYSTEM OPERATIONAL ***", 0xFFFFFFFF, 0x00000000);
    font_draw_string(20, 616, "Framebuffer: WORKING | Resolution: 1024x768", 0xFFFFFFFF, 0x00000000);
    font_draw_string(20, 632, "Memory: OK | Graphics: OK | Text: OK", 0xFFFFFFFF, 0x00000000);

    return 0;
}

/*
 * Mark splash screen as complete
 */
int splash_finish(void) {
    if (splash_active) {
        splash_report_progress(100, "Boot complete");

        // Show "Boot complete" message for 0.25 seconds before finishing
        pit_poll_delay_ms(250);

        splash_active = 0;
        klog_printf(KLOG_INFO, "SPLASH: Boot splash screen complete\n");

        // Draw the graphics demo screen
        splash_draw_graphics_demo();
    }
    return 0;
}

/*
 * Update splash screen with loading progress
 */
int splash_update_progress(int progress, const char *message) {
    if (!framebuffer_is_initialized()) {
        return -1;
    }

    // Get screen dimensions
    uint32_t width = framebuffer_get_width();
    uint32_t height = framebuffer_get_height();
    int center_x = width / 2;
    int center_y = height / 2;

    // Clear previous message area
    graphics_draw_rect_filled(center_x - 150, center_y + 155, 300, 20, SPLASH_BG_COLOR);

    // Draw new message
    if (message) {
        font_draw_string(center_x - 70, center_y + 160, message, SPLASH_TEXT_COLOR, 0x00000000);
    }

    // Update progress bar
    int progress_bar_width = 300;
    int progress_bar_height = 20;
    int progress_bar_x = center_x - progress_bar_width / 2;
    int progress_bar_y = center_y + 200;

    splash_draw_progress_bar(progress_bar_x, progress_bar_y, progress_bar_width, progress_bar_height, progress);

    return 0;
}

/*
 * Clear splash screen
 */
int splash_clear(void) {
    if (!framebuffer_is_initialized()) {
        return -1;
    }

    // Clear screen to black
    framebuffer_clear(0x00000000);
    return 0;
}

/*
 * Show splash screen with simple delay
 */
/* Removed splash_show_with_delay: callers should invoke splash_show_boot_screen directly */

File: video/src/splash.rs
use core::ffi::c_char;

use spin::Mutex;

use crate::font::font_draw_string;
use crate::framebuffer;
use crate::graphics::{self, GraphicsResult};
use slopos_drivers::pit;
use slopos_drivers::video_bridge::VideoError;

const SPLASH_BG_COLOR: u32 = 0x0000_0000;
const SPLASH_TEXT_COLOR: u32 = 0xE6E6_E6FF;
const SPLASH_SUBTEXT_COLOR: u32 = 0x9A9A_9AFF;
const SPLASH_ACCENT_COLOR: u32 = 0x00C2_7FFF;
const SPLASH_PROGRESS_TRACK_COLOR: u32 = 0x1A1A_1AFF;
const SPLASH_PROGRESS_FRAME_COLOR: u32 = 0x2E2E_2EFF;

const SPLASH_PROGRESS_WIDTH: i32 = 260;
const SPLASH_PROGRESS_HEIGHT: i32 = 10;
const SPLASH_MESSAGE_WIDTH: i32 = 320;
const SPLASH_MESSAGE_HEIGHT: i32 = 18;

const TEXT_TITLE: &[u8] = b"SLOPOS\0";
const TEXT_SUBTITLE: &[u8] = b"Safe boot\0";
const TEXT_INIT: &[u8] = b"Starting services...\0";

struct SplashState {
    active: bool,
    progress: i32,
}

impl SplashState {
    const fn new() -> Self {
        Self {
            active: false,
            progress: 0,
        }
    }
}

static STATE: Mutex<SplashState> = Mutex::new(SplashState::new());

fn framebuffer_ready() -> bool {
    framebuffer::framebuffer_is_initialized() != 0
}

fn ensure_framebuffer_ready() -> GraphicsResult<()> {
    if framebuffer_ready() {
        Ok(())
    } else {
        Err(VideoError::NoFramebuffer)
    }
}

fn splash_draw_logo(center_x: i32, center_y: i32) -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;

    let ring_radius = 28;
    graphics::graphics_draw_circle_filled(center_x, center_y, ring_radius, SPLASH_ACCENT_COLOR)?;
    graphics::graphics_draw_circle_filled(
        center_x,
        center_y,
        ring_radius - 4,
        SPLASH_BG_COLOR,
    )?;
    graphics::graphics_draw_rect_filled(
        center_x - 40,
        center_y + ring_radius + 10,
        80,
        2,
        SPLASH_ACCENT_COLOR,
    )?;

    Ok(())
}

fn splash_draw_progress_bar(
    x: i32,
    y: i32,
    width: i32,
    height: i32,
    progress: i32,
) -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;

    graphics::graphics_draw_rect_filled(x, y, width, height, SPLASH_PROGRESS_TRACK_COLOR)?;
    graphics::graphics_draw_rect(
        x - 1,
        y - 1,
        width + 2,
        height + 2,
        SPLASH_PROGRESS_FRAME_COLOR,
    )?;

    if progress > 0 {
        let fill_width = (width * progress) / 100;
        graphics::graphics_draw_rect_filled(x, y, fill_width, height, SPLASH_ACCENT_COLOR)?;
    }

    Ok(())
}
pub fn splash_show_boot_screen() -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;

    let mut state = STATE.lock();
    framebuffer::framebuffer_clear(SPLASH_BG_COLOR);

    let width = framebuffer::framebuffer_get_width() as i32;
    let height = framebuffer::framebuffer_get_height() as i32;
    let center_x = width / 2;
    let center_y = height / 2;

    splash_draw_logo(center_x, center_y - 90)?;
    let title_x = center_x - ((TEXT_TITLE.len() as i32 - 1) * 8 / 2);
    if font_draw_string(
        title_x,
        center_y - 35,
        TEXT_TITLE.as_ptr() as *const c_char,
        SPLASH_TEXT_COLOR,
        0,
    ) != 0
    {
        return Err(VideoError::Invalid);
    }
    let subtitle_x = center_x - ((TEXT_SUBTITLE.len() as i32 - 1) * 8 / 2);
    if font_draw_string(
        subtitle_x,
        center_y - 15,
        TEXT_SUBTITLE.as_ptr() as *const c_char,
        SPLASH_SUBTEXT_COLOR,
        0,
    ) != 0
    {
        return Err(VideoError::Invalid);
    }
    let message_x = center_x - (SPLASH_MESSAGE_WIDTH / 2);
    if font_draw_string(
        message_x,
        center_y + 15,
        TEXT_INIT.as_ptr() as *const c_char,
        SPLASH_SUBTEXT_COLOR,
        0,
    ) != 0
    {
        return Err(VideoError::Invalid);
    }

    let progress_bar_x = center_x - SPLASH_PROGRESS_WIDTH / 2;
    let progress_bar_y = center_y + 40;
    splash_draw_progress_bar(
        progress_bar_x,
        progress_bar_y,
        SPLASH_PROGRESS_WIDTH,
        SPLASH_PROGRESS_HEIGHT,
        0,
    )?;

    state.active = true;
    state.progress = 0;
    Ok(())
}
pub fn splash_update_progress(progress: i32, message: *const c_char) -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;

    let width = framebuffer::framebuffer_get_width() as i32;
    let height = framebuffer::framebuffer_get_height() as i32;
    let center_x = width / 2;
    let center_y = height / 2;

    graphics::graphics_draw_rect_filled(
        center_x - (SPLASH_MESSAGE_WIDTH / 2),
        center_y + 15,
        SPLASH_MESSAGE_WIDTH,
        SPLASH_MESSAGE_HEIGHT,
        SPLASH_BG_COLOR,
    )?;
    if !message.is_null()
        && font_draw_string(
            center_x - (SPLASH_MESSAGE_WIDTH / 2),
            center_y + 15,
            message,
            SPLASH_SUBTEXT_COLOR,
            0,
        ) != 0
    {
        return Err(VideoError::Invalid);
    }

    let progress_bar_x = center_x - SPLASH_PROGRESS_WIDTH / 2;
    let progress_bar_y = center_y + 40;
    splash_draw_progress_bar(
        progress_bar_x,
        progress_bar_y,
        SPLASH_PROGRESS_WIDTH,
        SPLASH_PROGRESS_HEIGHT,
        progress,
    )?;
    Ok(())
}
pub fn splash_report_progress(progress: i32, message: *const c_char) -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;

    let mut state = STATE.lock();
    if !state.active {
        return Err(VideoError::Invalid);
    }

    state.progress = progress.min(100);
    splash_update_progress(state.progress, message)?;

    let delay_ms = if state.progress <= 20 {
        300
    } else if state.progress <= 40 {
        250
    } else if state.progress <= 60 {
        280
    } else if state.progress <= 80 {
        320
    } else if state.progress <= 95 {
        280
    } else {
        250
    };

    pit::pit_poll_delay_ms(delay_ms as u32);
    Ok(())
}
pub fn splash_finish() -> GraphicsResult<()> {
    let mut state = STATE.lock();
    if state.active {
        splash_report_progress(100, b"Boot complete\0".as_ptr() as *const c_char)?;
        pit::pit_poll_delay_ms(250);
        state.active = false;
    }
    Ok(())
}
pub fn splash_clear() -> GraphicsResult<()> {
    ensure_framebuffer_ready()?;
    framebuffer::framebuffer_clear(SPLASH_BG_COLOR);
    Ok(())
}
